File: suites/osprey-compute-kernels/reference/cases/exprtree/exprtree.c
/* Expression tree — build a heap-allocated tree, then evaluate it. */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define MINSTD 16807
#define MODULUS 2147483647
#define BIG_MOD 1000000007

#define TAG_LIT 0
#define TAG_ADD 1
#define TAG_MUL 2

static int64_t r1(int64_t x) { return (x * MINSTD) % MODULUS; }

static int64_t hash_at(int64_t s, int64_t i) { return r1(r1(s + i)); }

static int64_t read_seed(void) {
    char line[256];
    int64_t m = 0;
    if (fgets(line, sizeof line, stdin) != NULL) {
        char *end = NULL;
        long long parsed = strtoll(line, &end, 10);
        if (end != line) {
            m = (int64_t)parsed;
        }
    }
    if (m == 0) {
        return 1;
    }
    uint64_t val = 0;
    FILE *f = fopen("/dev/urandom", "rb");
    if (f != NULL) {
        if (fread(&val, sizeof val, 1, f) != 1) {
            val = 0;
        }
        fclose(f);
    }
    return (int64_t)(val % 2147483646) + 1;
}

struct Expr {
    int tag;
    int64_t v;
    struct Expr *l, *r;
};

static struct Expr *build(int64_t s, int64_t idx, int64_t depth) {
    struct Expr *e = malloc(sizeof *e);
    if (e == NULL) {
        return NULL;
    }
    if (depth == 0) {
        e->tag = TAG_LIT;
        e->v = hash_at(s, idx) % 1000;
        e->l = NULL;
        e->r = NULL;
    } else {
        int64_t op = hash_at(s, idx) % 2;
        e->v = 0;
        e->l = build(s, idx * 2 + 1, depth - 1);
        e->r = build(s, idx * 2 + 2, depth - 1);
        e->tag = op == 0 ? TAG_ADD : TAG_MUL;
    }
    return e;
}

static int64_t eval(const struct Expr *e) {
    switch (e->tag) {
        case TAG_LIT:
            return e->v;
        case TAG_ADD:
            return (eval(e->l) + eval(e->r)) % BIG_MOD;
        default:
            return (eval(e->l) * eval(e->r)) % BIG_MOD;
    }
}

int main(void) {
    int64_t seed = read_seed();

    int64_t acc = 0;
    for (int64_t t = 0; t < 10; t++) {
        acc = (acc + eval(build(seed + t * 7, 1, 14))) % BIG_MOD;
    }

    printf("%lld\n", (long long)acc);
    return 0;
}

File: suites/osprey-compute-kernels/reference/cases/exprtree/exprtree.hs
{-# LANGUAGE BangPatterns #-}

module Main (main) where

import Control.Monad (replicateM)
import Data.Char (isSpace)
import Data.List (foldl')
import System.IO
  ( IOMode (ReadMode)
  , hGetChar
  , isEOF
  , withBinaryFile
  )
import Text.Read (readMaybe)

minstd :: Int
minstd = 16807

modulus :: Int
modulus = 2147483647

bigMod :: Int
bigMod = 1000000007

r1 :: Int -> Int
r1 x = (x * minstd) `mod` modulus

hashAt :: Int -> Int -> Int
hashAt s i = r1 (r1 (s + i))

readFirstLine :: IO String
readFirstLine = do
  eof <- isEOF
  if eof then pure "" else getLine

readSeed :: IO Int
readSeed = do
  line <- readFirstLine
  let m = maybe 0 id (readMaybe (filter (not . isSpace) line))
  if m == 0
    then pure 1
    else do
      cs <- withBinaryFile "/dev/urandom" ReadMode (replicateM 8 . hGetChar)
      let val = foldl (\acc c -> acc * 256 + fromEnum c) 0 cs
      pure (abs val `mod` 2147483646 + 1)

data Expr = Lit !Int | Add Expr Expr | Mul Expr Expr

build :: Int -> Int -> Int -> Expr
build s idx depth
  | depth == 0 = Lit (hashAt s idx `mod` 1000)
  | otherwise =
      let op = hashAt s idx `mod` 2
          l = build s (idx * 2 + 1) (depth - 1)
          r = build s (idx * 2 + 2) (depth - 1)
       in if op == 0 then Add l r else Mul l r

eval :: Expr -> Int
eval (Lit v) = v
eval (Add l r) = (eval l + eval r) `mod` bigMod
eval (Mul l r) = (eval l * eval r) `mod` bigMod

main :: IO ()
main = do
  seed <- readSeed
  let acc =
        foldl'
          (\ !a t -> (a + eval (build (seed + t * 7) 1 14)) `mod` bigMod)
          0
          [0 .. 9]
  print acc
